### PR TITLE
[DataCatalog2.0]: Move pattern resolution logic - project cli

### PR DESCRIPTION
**Pull request by ElenaKhaustova**
## Description
This PR is a part of https://github.com/kedro-org/kedro/issues/4110 and is done on top of https://github.com/kedro-org/kedro/pull/4123

It serves as an example of changes required to keep both implementations together so that users can switch between them and use both, see the related [comment](https://github.com/kedro-org/kedro/issues/3995#issuecomment-2286943456)

**For the reviewers:** this PR does not include tests, please see the suggested order of work in [this comment](https://github.com/kedro-org/kedro/issues/3995#issuecomment-2319079315)

## Development notes
This PR includes modification of `context`, `session`, `runners` and `kedro run` command to be compatible with `KedroDataCatalog` and `DataCatalogConfigResolver`.

Currently, one can use old `DataCatalog` as usual without any changes or use `KedroDataCatalog` with `kedro run` by adding  the following flag:

```
kedro run --new_catalog
```

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/framework/cli/project.py [hunk 1]
```diff
@@ -199,6 +200,7 @@ def package(metadata: ProjectMetadata) -> None:
     help=PARAMS_ARG_HELP,
     callback=_split_params,
 )
+@click.option("--new_catalog", "new_catalog", is_flag=True, help=NEW_CATALOG_ARG_HELP)
```
Comment (astrojuanlu) [inline kedro/framework/cli/project.py:203]: How does this flag play along with the `DATA_CATALOG_CLASS` defined in `settings.py`?

https://github.com/kedro-org/kedro/blob/7a16e1a48a2b1433a39b334787c534569abc4dbe/kedro/templates/project/%7B%7B%20cookiecutter.repo_name%20%7D%7D/src/%7B%7B%20cookiecutter.python_package%20%7D%7D/settings.py#L44-L46

Comment (idanov) [reply kedro/framework/cli/project.py:203]: Is this suggestion just for the PoC or to be included in the final code? Adding a temporary flag just for the new catalog has a very low API surface to benefit ratio. Removing ANY flag from the CLI is a breaking change, including this. Could we just utilise the existing methodology as pointed by @astrojuanlu ?

File: kedro/runner/runner.py [hunk 2]
```diff
@@ -83,7 +83,9 @@ def run(
         """
 
         hook_or_null_manager = hook_manager or _NullPluginManager()
-        catalog = catalog.shallow_copy()
+
+        # Seems like it's not needed even for DataCatalog
+        # catalog = catalog.shallow_copy()
```
Comment (idanov) [inline kedro/runner/runner.py:88]: Will that stay in? Or just for the PoC?

File: kedro/runner/runner.py [hunk 3]
```diff
@@ -99,9 +101,13 @@ def run(
             )
 
         # Identify MemoryDataset in the catalog
+        catalog_datasets = (
```
Comment (idanov) [inline kedro/runner/runner.py:104]: That's exactly the kind of thing an abstract DataCatalog (or a Protocol to this end) will prevent. Having code to check what type of instance we are working with is a bit of an anti-pattern in most cases (bar some exceptional cases emulating pattern matching). Could we somehow create a minimal interface which is used within Kedro, give it a name and program against it?

File: kedro/runner/runner.py [hunk 4]
```diff
@@ -380,7 +398,7 @@ def _find_initial_node_group(pipeline: Pipeline, nodes: Iterable[Node]) -> list[
 
 def run_node(
     node: Node,
-    catalog: DataCatalog,
+    catalog: DataCatalog | KedroDataCatalog,
```
Comment (idanov) [inline kedro/runner/runner.py:401]: If this is for the PoC, could we (just temporarily) come up with something like:
```python
type AbstractDataCatalog = DataCatalog | KedroDataCatalog
```

And then simply use that everywhere until we design the `AbstractDataCatalog`?

**Review by idanov — CHANGES_REQUESTED** [review]

File: kedro/framework/cli/project.py (hunk 1 above)
Comment (ElenaKhaustova) [reply kedro/framework/cli/project.py:203]: I did this temporarily to simplify running the project with both catalogs without modifying `DATA_CATALOG_CLASS` for the project before running it.  This flag and `DATA_CATALOG_CLASS_NEW` will go away later on.

File: kedro/runner/runner.py (hunk 2 above)
Comment (ElenaKhaustova) [reply kedro/runner/runner.py:88]: I should go IMO as it doesn't seem useful, but I was going to double-check why it was there

File: kedro/runner/runner.py [hunk 5]
```diff
@@ -110,9 +116,10 @@ def run(
         free_outputs = pipeline.outputs() - (set(registered_ds) - memory_datasets)
 
         # Register the default dataset pattern with the catalog
```
Comment (merelcht) [inline kedro/runner/runner.py:118]: How does this work for the new Catalog?

Comment (ElenaKhaustova) [reply kedro/runner/runner.py:118]: All the patterns are managed by `DataCatalogConfigResolver` which has `dataset_patterns`, `default_pattern` and `runtime_patterns` and stores them explicitly - https://github.com/kedro-org/kedro/blob/506470aa2386d7d7d5209cde9b53e09e47c4dcf3/kedro/io/catalog_config_resolver.py#L98

Previously we used shallow copy just to add `runtime_patterns` (line that you're pointing to).  Now we have a dedicated method for that - https://github.com/kedro-org/kedro/blob/506470aa2386d7d7d5209cde9b53e09e47c4dcf3/kedro/io/catalog_config_resolver.py#L246

We add runtime patterns to the `DataCatalogConfigResolver` at the level of the session where runner is already initialised, so runner does not depend on `DataCatalogConfigResolver`.
https://github.com/kedro-org/kedro/blob/506470aa2386d7d7d5209cde9b53e09e47c4dcf3/kedro/framework/session/session.py#L409

**Comment by ElenaKhaustova** [comment]
Replaced with https://github.com/kedro-org/kedro/pull/4123 and https://github.com/kedro-org/kedro/pull/4151

**Pull request closed without merging**